Add some CoreText stub definitions

jlucfarias opened the pull request:
I've tried to add all constants of CTFont (unfortunatelfy I do not have time to add constants and stubs for other files) but I've keeped constants unused at this time commented to be uncommented later.

One suggestion is to move all frameworks from here (and other darling projects) to darling repo (or a darling specific repo for frameworks) to keep only one place with all frameworks. IMHO this is a more maintainable action to get darling project more active.
A lot of definitions can be get from the abandoned project [darling-coregraphics](https://github.com/darlinghq/darling-coregraphics/tree/master) and I used for reference for this changes.

Fixes [darling#797](https://github.com/darlinghq/darling/issues/797)
Fixes [darling#798](https://github.com/darlinghq/darling/issues/798)
Fixes [darling#1527](https://github.com/darlinghq/darling/issues/1527)

CuriousTommy commented on CoreText/CTFont.m:
```diff
@@ -18,10 +18,53 @@ this software and associated documentation files (the "Software"), to deal in
+// const CFStringRef kCTFontCopyrightNameKey = @"kCTFontCopyrightNameKey";
+// const CFStringRef kCTFontFamilyNameKey = @"kCTFontFamilyNameKey";
+// const CFStringRef kCTFontSubFamilyNameKey = @"kCTFontSubFamilyNameKey";
+const CFStringRef kCTFontStyleNameKey = @"kCTFontStyleNameKey";
```
Please don't change the old key value unless you can verify on a macOS machine that the new value is the correct one. I did a test on my Mac machine and the old value was the correct one.

CuriousTommy commented on CoreText/CTFont.m:
```diff
@@ -18,10 +18,53 @@ this software and associated documentation files (the "Software"), to deal in
+// const CFStringRef kCTFontPostScriptCIDNameKey =
+// @"kCTFontPostScriptCIDNameKey";
+
+// const CFStringRef kCTFontVariationAxisIdentifierKey =
+// @"kCTFontVariationAxisIdentifierKey"; const CFStringRef
+// kCTFontVariationAxisMinimumValueKey = @"kCTFontVariationAxisMinimumValueKey";
+// const CFStringRef kCTFontVariationAxisMaximumValueKey =
+// @"kCTFontVariationAxisMaximumValueKey"; const CFStringRef
+// kCTFontVariationAxisDefaultValueKey = @"kCTFontVariationAxisDefaultValueKey";
+// const CFStringRef kCTFontVariationAxisNameKey =
+// @"kCTFontVariationAxisNameKey"; const CFStringRef
+// kCTFontVariationAxisHiddenKey = @"kCTFontVariationAxisHiddenKey";
+
+// const CFStringRef kCTFontFeatureTypeIdentifierKey =
+// @"kCTFontFeatureTypeIdentifierKey"; const CFStringRef
+// kCTFontFeatureTypeNameKey = @"kCTFontFeatureTypeNameKey"; const CFStringRef
+// kCTFontFeatureTypeExclusiveKey = @"kCTFontFeatureTypeExclusiveKey"; const
+// CFStringRef kCTFontFeatureTypeSelectorsKey =
+// @"kCTFontFeatureTypeSelectorsKey"; const CFStringRef
+// kCTFontFeatureSelectorIdentifierKey = @"kCTFontFeatureSelectorIdentifierKey";
+// const CFStringRef kCTFontFeatureSelectorNameKey =
+// @"kCTFontFeatureSelectorNameKey"; const CFStringRef
+// kCTFontFeatureSelectorDefaultKey = @"kCTFontFeatureSelectorDefaultKey"; const
+// CFStringRef kCTFontFeatureSelectorSettingKey =
+// @"kCTFontFeatureSelectorSettingKey"; const CFStringRef
+// kCTFontFeatureSampleTextKey = @"kCTFontFeatureSampleTextKey"; const
+// CFStringRef kCTFontFeatureTooltipTextKey = @"kCTFontFeatureTooltipTextKey";
```
Are you using an auto formatting tool? It seems to be messing up the formatting of the commented out code...

CuriousTommy commented on CoreText/CTFont.m:
```diff
@@ -114,39 +157,39 @@ CFArrayRef CTFontCopyAvailableTables(CTFontRef font, CTFontTableOptions options)
 }
 
-CTFontRef CTFontCreateCopyWithAttributes(CTFontRef font, CGFloat size, const CGAffineTransform *matrix, CTFontDescriptorRef attributes)
+CTFontRef CTFontCreateCopyWithAttributes(CTFontRef font, CGFloat size,
+                                         const CGAffineTransform *matrix,
+                                         CTFontDescriptorRef attributes)
```
I have mixed feelings about these formatting changes... I rather keep them the way they currently are.

CuriousTommy commented on CoreText/CTFontDescriptor.m:
```diff
@@ -2,7 +2,35 @@
-CFTypeRef CTFontDescriptorCopyAttribute(CTFontDescriptorRef descriptor, CFStringRef attribute)
+// const CFStringRef kCTFontURLAttribute = @"kCTFontURLAttribute";
+// const CFStringRef kCTFontNameAttribute = @"kCTFontNameAttribute";
+// const CFStringRef kCTFontDisplayNameAttribute =
+// @"kCTFontDisplayNameAttribute"; const CFStringRef kCTFontFamilyNameAttribute
+// = @"kCTFontFamilyNameAttribute"; const CFStringRef kCTFontStyleNameAttribute
+// = @"kCTFontStyleNameAttribute"; const CFStringRef kCTFontTraitsAttribute =
+// @"kCTFontTraitsAttribute"; const CFStringRef kCTFontVariationAttribute =
+// @"kCTFontVariationAttribute"; const CFStringRef kCTFontSizeAttribute =
+// @"kCTFontSizeAttribute"; const CFStringRef kCTFontMatrixAttribute =
+// @"kCTFontMatrixAttribute"; const CFStringRef kCTFontCascadeListAttribute =
+// @"kCTFontCascadeListAttribute"; const CFStringRef
+// kCTFontCharacterSetAttribute = @"kCTFontCharacterSetAttribute"; const
+// CFStringRef kCTFontLanguagesAttribute = @"kCTFontLanguagesAttribute"; const
+// CFStringRef kCTFontBaselineAdjustAttribute =
+// @"kCTFontBaselineAdjustAttribute"; const CFStringRef
+// kCTFontMacintoshEncodingsAttribute = @"kCTFontMacintoshEncodingsAttribute";
+const CFStringRef kCTFontFeaturesAttribute = @"kCTFontFeaturesAttribute";
+// const CFStringRef kCTFontFeatureSettingsAttribute =
+// @"kCTFontFeatureSettingsAttribute"; const CFStringRef
+// kCTFontFixedAdvanceAttribute = @"kCTFontFixedAdvanceAttribute"; const
+// CFStringRef kCTFontOrientationAttribute = @"kCTFontOrientationAttribute";
+// const CFStringRef kCTFontEnabledAttribute = @"kCTFontEnabledAttribute";
+// const CFStringRef kCTFontFormatAttribute = @"kCTFontFormatAttribute";
+// const CFStringRef kCTFontRegistrationScopeAttribute =
+// @"kCTFontRegistrationScopeAttribute"; const CFStringRef
+// kCTFontPriorityAttribute = @"kCTFontPriorityAttribute";
```
Same here, looks like some weird formatting is going on here...

CuriousTommy commented on CoreText/constants.c:
```diff
@@ -48,7 +48,6 @@ const CFStringRef kCTFontCharacterSetAttribute = CFSTR("NSFontCharacterSetAttrib
 const CFStringRef kCTFontLanguagesAttribute = CFSTR("NSFontLanguagesAttribute");
 const CFStringRef kCTFontBaselineAdjustAttribute = CFSTR("NSFontBaselineAdjustAttribute");
 const CFStringRef kCTFontMacintoshEncodingsAttribute = CFSTR("NSFontMacintoshEncodingsAttribute");
-const CFStringRef kCTFontFeaturesAttribute = CFSTR("NSFontFeaturesAttribute");
```
It's seems odd to me that you move this value, but you didn't move the other values (instead you made commented out copies of the other keys...)

I don't like the way you are handing this... you should either move these keys to the new location (uncommented), or leave them here.

CuriousTommy commented on CoreText/CTFont.m:
```diff
@@ -18,10 +18,53 @@ this software and associated documentation files (the "Software"), to deal in
+#import <CoreText/KTFont.h>
 
-const CFStringRef kCTFontStyleNameKey = "CTFontSubFamilyName";
+// const CFStringRef kCTFontCopyrightNameKey = @"kCTFontCopyrightNameKey";
+// const CFStringRef kCTFontFamilyNameKey = @"kCTFontFamilyNameKey";
+// const CFStringRef kCTFontSubFamilyNameKey = @"kCTFontSubFamilyNameKey";
+const CFStringRef kCTFontStyleNameKey = @"kCTFontStyleNameKey";
+// const CFStringRef kCTFontUniqueNameKey = @"kCTFontUniqueNameKey";
+const CFStringRef kCTFontFullNameKey = @"kCTFontFullNameKey";
+// const CFStringRef kCTFontVersionNameKey = @"kCTFontVersionNameKey";
+// const CFStringRef kCTFontPostScriptNameKey = @"kCTFontPostScriptNameKey";
+// const CFStringRef kCTFontTrademarkNameKey = @"kCTFontTrademarkNameKey";
+// const CFStringRef kCTFontManufacturerNameKey = @"kCTFontManufacturerNameKey";
+// const CFStringRef kCTFontDesignerNameKey = @"kCTFontDesignerNameKey";
+// const CFStringRef kCTFontDescriptionNameKey = @"kCTFontDescriptionNameKey";
+// const CFStringRef kCTFontVendorURLNameKey = @"kCTFontVendorURLNameKey";
+// const CFStringRef kCTFontDesignerURLNameKey = @"kCTFontDesignerURLNameKey";
+// const CFStringRef kCTFontLicenseNameKey = @"kCTFontLicenseNameKey";
+// const CFStringRef kCTFontLicenseURLNameKey = @"kCTFontLicenseURLNameKey";
+// const CFStringRef kCTFontSampleTextNameKey = @"kCTFontSampleTextNameKey";
+// const CFStringRef kCTFontPostScriptCIDNameKey =
+// @"kCTFontPostScriptCIDNameKey";
```
For the other values, here are the key values that macOS reports:

```suggestion
// const CFStringRef kCTFontCopyrightNameKey = @"CTFontCopyrightName";
// const CFStringRef kCTFontFamilyNameKey = @"CTFontFamilyName";
// const CFStringRef kCTFontSubFamilyNameKey = @"CTFontSubFamilyName";
const CFStringRef kCTFontStyleNameKey = @"CTFontSubFamilyName";
// const CFStringRef kCTFontUniqueNameKey = @"CTFontUniqueName";
const CFStringRef kCTFontFullNameKey = @"CTFontFullName";
// const CFStringRef kCTFontVersionNameKey = @"CTFontVersionName";
// const CFStringRef kCTFontPostScriptNameKey = @"CTFontPostScriptName";
// const CFStringRef kCTFontTrademarkNameKey = @"CTFontTrademarkName";
// const CFStringRef kCTFontManufacturerNameKey = @"CTFontManufacturerName";
// const CFStringRef kCTFontDesignerNameKey = @"CTFontDesignerName";
// const CFStringRef kCTFontDescriptionNameKey = @"CTFontDescriptionName";
// const CFStringRef kCTFontVendorURLNameKey = @"CTFontVendorURLName";
// const CFStringRef kCTFontDesignerURLNameKey = @"CTFontDesignerURLName";
// const CFStringRef kCTFontLicenseNameKey = @"CTFontLicenseNameName";
// const CFStringRef kCTFontLicenseURLNameKey = @"CTFontLicenseURLName";
// const CFStringRef kCTFontSampleTextNameKey = @"CTFontSampleTextName";
// const CFStringRef kCTFontPostScriptCIDNameKey = @"CTFontPostScriptCIDName";
```

CuriousTommy reviewed (changes requested):
There are some changes I need you to fix first before I can merge this.

On a side note, is there a particular reason why you want to keep most of the new changes commented out? I think it would be better if we left them un-commented.

jlucfarias commented:
@CuriousTommy I think I fixed all requested changes: reverted the formatting changes, uncomment constants and moved from constants.c to the new file.
My reasons for keep commented was that it is not used right now to issues I'm trying to fix, but I wanted to add these constants to be available to anyone that want to contribute with the project in the future

CuriousTommy commented on CoreText/include/CoreText/CTFont.h:
```diff
@@ -16,91 +16,280 @@ FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+CORETEXT_EXPORT CTFontRef CTFontCreateForStringWithLanguage(CTFontRef currentFont, CFStringRef string, CFRange range, CFStringRef language);
+
+CORETEXT_EXPORT CTFontDescriptorRef CTFontCopyFontDescriptor(CTFontRef font);
+CORETEXT_EXPORT CFTypeRef CTFontCopyAttribute(CTFontRef font,
+                                              CFStringRef attribute);
```
This should be one line, not two
```suggestion
CORETEXT_EXPORT CFTypeRef CTFontCopyAttribute(CTFontRef font, CFStringRef attribute);
```

CuriousTommy commented on CoreText/include/CoreText/CTFont.h:
```diff
@@ -16,91 +16,280 @@ FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+
+CORETEXT_EXPORT CGFontRef CTFontCopyGraphicsFont(
+        CTFontRef font, CTFontDescriptorRef _Nullable *attributes);
+CORETEXT_EXPORT CTFontRef CTFontCreateWithGraphicsFont(
+        CGFontRef cgFont, CGFloat size, CGAffineTransform *xform,
+        /* CTFontDescriptorRef */ void *attributes);
```
Same here (Use `CTFontDescriptorRef` instead of `void*`)

CuriousTommy commented on CoreText/include/CoreText/CTFont.h:
```diff
@@ -16,91 +16,280 @@ FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+CORETEXT_EXPORT CFArrayRef CTFontCopyFeatureSettings(CTFontRef font);
 
-CORETEXT_EXPORT CFStringRef CTFontCopyName(CTFontRef font, CFStringRef nameKey);
+CORETEXT_EXPORT bool CTFontGetGlyphsForCharacters(CTFontRef self,
+                                                  const UniChar *characters,
+                                                  CGGlyph *glyphs,
+                                                  size_t count);
```
```suggestion
CORETEXT_EXPORT bool CTFontGetGlyphsForCharacters(CTFontRef font, const UniChar *characters, CGGlyph *glyphs, CFIndex count);
```

CuriousTommy commented on CoreText/include/CoreText/CTFont.h:
```diff
@@ -16,91 +16,280 @@ FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
                                                    CGAffineTransform *xform);
+CORETEXT_EXPORT CGGlyph CTFontGetGlyphWithName(CTFontRef font, CFStringRef glyphName);
+CORETEXT_EXPORT CGRect CTFontGetBoundingRectsForGlyphs(CTFontRef font, CTFontOrientation orientation, const CGGlyph *glyphs, CGRect *boundingRects, CFIndex count);
+CORETEXT_EXPORT void CTFontGetAdvancesForGlyphs(CTFontRef self, int orientation,
+                                                const CGGlyph *glyphs,
+                                                CGSize *advances, size_t count);
```
```suggestion
CORETEXT_EXPORT double CTFontGetAdvancesForGlyphs(CTFontRef font, CTFontOrientation orientation,
                                                const CGGlyph *glyphs,
                                                CGSize *advances, CFIndex count);
```

CuriousTommy reviewed (changes requested):
Sorry for taking so long to review your new changes. Been feeling burned out lately...

CuriousTommy commented:
@jlucfarias Did you get a chance to build and run your changes? I'm getting some errors:

```
/home/thomasa/Downloads/darling2/framework-include/CoreText/CTFont.h:189:14: error: redefinition of 'CTFontTableOptions'
  189 | typedef enum CTFontTableOptions : uint32_t {
      |              ^
/home/thomasa/Downloads/darling2/framework-include/CoreText/CTFont.h:124:30: note: previous definition is here
  124 | typedef CF_OPTIONS(uint32_t, CTFontTableOptions) {
      |                              ^
/home/thomasa/Downloads/darling2/framework-include/CoreText/CTFont.h:190:5: error: redefinition of enumerator 'kCTFontTableOptionNoOptions'
  190 |     kCTFontTableOptionNoOptions = 0,
      |     ^
/home/thomasa/Downloads/darling2/framework-include/CoreText/CTFont.h:125:5: note: previous definition is here
  125 |     kCTFontTableOptionNoOptions        = 0,
      |     ^
/home/thomasa/Downloads/darling2/framework-include/CoreText/CTFont.h:191:5: error: redefinition of enumerator 'kCTFontTableOptionExcludeSynthetic'
  191 |     kCTFontTableOptionExcludeSynthetic = (1 << 0) // Deprecated
      |     ^
/home/thomasa/Downloads/darling2/framework-include/CoreText/CTFont.h:126:5: note: previous definition is here
  126 |     kCTFontTableOptionExcludeSynthetic = 1,
      |     ^
/home/thomasa/Downloads/darling2/framework-include/CoreText/CTFont.h:192:3: error: typedef redefinition with different types ('enum (unnamed enum at /home/thomasa/Downloads/darling2/framework-include/CoreText/CTFont.h:189:14)' vs 'enum CTFontTableOptions')
  192 | } CTFontTableOptions;
      |   ^
/home/thomasa/Downloads/darling2/framework-include/CoreText/CTFont.h:124:30: note: previous definition is here
  124 | typedef CF_OPTIONS(uint32_t, CTFontTableOptions) {
      |                              ^
/home/thomasa/Downloads/darling2/framework-include/CoreText/CTFont.h:219:125: error: unknown type name 'CTFontSymbolicTraits'
  219 | CORETEXT_EXPORT CTFontRef CTFontCreateCopyWithSymbolicTraits(CTFontRef font, CGFloat size, const CGAffineTransform *matrix, CTFontSymbolicTraits symTraitValue, CTFontSymbolicTraits symTraitMask);
      |                                                                                                                             ^
/home/thomasa/Downloads/darling2/framework-include/CoreText/CTFont.h:219:161: error: unknown type name 'CTFontSymbolicTraits'
  219 | CORETEXT_EXPORT CTFontRef CTFontCreateCopyWithSymbolicTraits(CTFontRef font, CGFloat size, const CGAffineTransform *matrix, CTFontSymbolicTraits symTraitValue, CTFontSymbolicTraits symTraitMask);
      |                                                                                                                                                                 ^
/home/thomasa/Downloads/darling2/framework-include/CoreText/CTFont.h:228:17: error: unknown type name 'CTFontSymbolicTraits'
  228 | CORETEXT_EXPORT CTFontSymbolicTraits CTFontGetSymbolicTraits(CTFontRef font);
      |                 ^
/home/thomasa/Downloads/darling2/framework-include/CoreText/CTFont.h:260:72: error: unknown type name 'CTFontOrientation'
  260 | CORETEXT_EXPORT CGRect CTFontGetBoundingRectsForGlyphs(CTFontRef font, CTFontOrientation orientation, const CGGlyph *glyphs, CGRect *boundingRects, CFIndex count);
      |                                                                        ^
/home/thomasa/Downloads/darling2/framework-include/CoreText/CTFont.h:261:67: error: unknown type name 'CTFontOrientation'
  261 | CORETEXT_EXPORT double CTFontGetAdvancesForGlyphs(CTFontRef font, CTFontOrientation orientation,
      |                                                                   ^
/home/thomasa/Downloads/darling2/framework-include/CoreText/CTFont.h:282:17: error: unknown type name 'ATSFontRef'; did you mean 'CTFontRef'?
  282 | CORETEXT_EXPORT ATSFontRef CTFontGetPlatformFont(CTFontRef font, CTFontDescriptorRef  _Nullable *attributes); // Deprecated
      |                 ^~~~~~~~~~
      |                 CTFontRef
/home/thomasa/Downloads/darling2/framework-include/CoreText/CTFont.h:27:44: note: 'CTFontRef' declared here
   27 | typedef struct CF_BRIDGED_TYPE(id) KTFont *CTFontRef;
      |                                            ^
/home/thomasa/Downloads/darling2/framework-include/CoreText/CTFont.h:283:56: error: unknown type name 'ATSFontRef'; did you mean 'CTFontRef'?
  283 | CORETEXT_EXPORT CTFontRef CTFontCreateWithPlatformFont(ATSFontRef platformFont, CGFloat size, const CGAffineTransform *matrix, CTFontDescriptorRef attributes); // Deprecated
      |                                                        ^~~~~~~~~~
      |                                                        CTFontRef
/home/thomasa/Downloads/darling2/framework-include/CoreText/CTFont.h:27:44: note: 'CTFontRef' declared here
   27 | typedef struct CF_BRIDGED_TYPE(id) KTFont *CTFontRef;
      |                                            ^
3 warnings and 11 errors generated.
```

jlucfarias commented:
@CuriousTommy sorry for the late response. I couldn't build at the moment I sent my changes, but now I builded and fixed all errors related to Coretext changes. Please, try again

CuriousTommy commented on CoreText/include/CoreText/CTFont.h:
```diff
@@ -16,91 +16,277 @@ FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 
-CORETEXT_EXPORT CFStringRef CTFontCopyName(CTFontRef font, CFStringRef nameKey);
+CORETEXT_EXPORT bool CTFontGetGlyphsForCharacters(CTFontRef font, const UniChar *characters, CGGlyph *glyphs, CFIndex count);
+CORETEXT_EXPORT void CTFontDrawGlyphs(CTFontRef font, const CGGlyph *glyphs, const CGPoint *positions, size_t count, CGContextRef context);
```
Don't forget to include the header file for [`CGContextRef`](https://developer.apple.com/documentation/coregraphics/cgcontextref)

```
In file included from /home/[user]/Downloads/darling2/src/external/cocotron/CoreText/CTFont.m:20:
/home/[user]/Downloads/darling2/framework-include/CoreText/CTFont.h:269:118: error: unknown type name 'CGContextRef'; did you mean 'CGFontRef'?
  269 | CORETEXT_EXPORT void CTFontDrawGlyphs(CTFontRef font, const CGGlyph *glyphs, const CGPoint *positions, size_t count, CGContextRef context);
      |                                                                                                                      ^~~~~~~~~~~~
      |                                                                                                                      CGFontRef
```

CuriousTommy commented on CoreText/CTFont.m:
```diff
@@ -39,43 +95,134 @@ CTFontRef CTFontCreateUIFontForLanguage(CTFontUIFontType uiFontType,
+CGAffineTransform CTFontGetMatrix(CTFontRef font)
+{
+    printf("STUB %s\n", __PRETTY_FUNCTION__);
+    return nil;
```
One thing to keep in mind is that not all return values are pointer based, so `nil` would not work here.

[`CGAffineTransform`](https://developer.apple.com/documentation/corefoundation/cgaffinetransform?language=objc)

```
/home/[user]/Downloads/darling2/src/external/cocotron/CoreText/CTFont.m:154:12: error: returning 'void *' from a function with incompatible result type 'CGAffineTransform'
  154 |     return nil;
      |            ^~~
```

CuriousTommy commented on CoreText/CTFont.m:
```diff
@@ -39,43 +95,134 @@ CTFontRef CTFontCreateUIFontForLanguage(CTFontUIFontType uiFontType,
-    // FIXME: change getGlyphs: to return a BOOL
-    return YES;
+    printf("STUB %s\n", __PRETTY_FUNCTION__);
+    return nil;
```
Same here [`CTFontSymbolicTraits`](https://developer.apple.com/documentation/coretext/ctfontsymbolictraits?language=objc)

```
/home/[user]/Downloads/darling2/src/external/cocotron/CoreText/CTFont.m:160:12: error: incompatible pointer to integer conversion returning 'void *' from a function with result type 'CTFontSymbolicTraits' (aka 'enum CTFontSymbolicTraits') [-Wint-conversion]
  160 |     return nil;
      |
```

CuriousTommy commented on CoreText/CTFont.m:
```diff
@@ -39,43 +95,134 @@ CTFontRef CTFontCreateUIFontForLanguage(CTFontUIFontType uiFontType,
+CFStringEncoding CTFontGetStringEncoding(CTFontRef font)
+{
+    printf("STUB %s\n", __PRETTY_FUNCTION__);
+    return nil;
```
Same here [`CFStringEncoding`](https://developer.apple.com/documentation/corefoundation/cfstringencoding?language=objc)

```
/home/[user]/Downloads/darling2/src/external/cocotron/CoreText/CTFont.m:219:12: error: incompatible pointer to integer conversion returning 'void *' from a function with result type 'CFStringEncoding' (aka 'unsigned int') [-Wint-conversion]
  219 |     return nil;
      |
```

CuriousTommy commented on CoreText/CTFont.m:
```diff
@@ -90,31 +237,129 @@ CGFloat CTFontGetLeading(CTFontRef self) {
+unsigned int CTFontGetUnitsPerEm(CTFontRef font)
+{
+    printf("STUB %s\n", __PRETTY_FUNCTION__);
+    return nil;
```
Same here

```
/home/[user]/Downloads/darling2/src/external/cocotron/CoreText/CTFont.m:243:12: error: incompatible pointer to integer conversion returning 'void *' from a function with result type 'unsigned int' [-Wint-conversion]
  243 |     return nil;
      |
```

CuriousTommy commented on CoreText/CTFont.m:
```diff
@@ -90,31 +237,129 @@ CGFloat CTFontGetLeading(CTFontRef self) {
+                                       CFIndex count)
+{
+    printf("STUB %s\n", __PRETTY_FUNCTION__);
+    return nil;
```
Same here [`CGRect`](https://developer.apple.com/documentation/corefoundation/cgrect?language=objc)

```
/home/[user]/Downloads/darling2/src/external/cocotron/CoreText/CTFont.m:291:12: error: returning 'void *' from a function with incompatible result type 'CGRect' (aka 'struct CGRect')
  291 |     return nil;
      |
```

CuriousTommy commented on CoreText/CTFont.m:
```diff
@@ -90,31 +237,129 @@ CGFloat CTFontGetLeading(CTFontRef self) {
+                                       CFOptionFlags options)
+{
+    printf("STUB %s\n", __PRETTY_FUNCTION__);
+    return nil;
```
Same here [`CGRect`](https://developer.apple.com/documentation/corefoundation/cgrect?language=objc)

```
/home/[user]/Downloads/darling2/src/external/cocotron/CoreText/CTFont.m:314:12: error: returning 'void *' from a function with incompatible result type 'CGRect' (aka 'struct CGRect')
  314 |     return nil;
      |
```

CuriousTommy commented on CoreText/CTFont.m:
```diff
@@ -90,31 +237,129 @@ CGFloat CTFontGetLeading(CTFontRef self) {
+CGGlyph CTFontGetGlyphWithName(CTFontRef font, CFStringRef glyphName)
 {
     printf("STUB %s\n", __PRETTY_FUNCTION__);
     return nil;
```
Same here [`CGGlyph`](https://developer.apple.com/documentation/coregraphics/cgglyph?language=objc)

```
/home/[user]/Downloads/darling2/src/external/cocotron/CoreText/CTFont.m:283:12: error: incompatible pointer to integer conversion returning 'void *' from a function with result type 'CGGlyph' (aka 'unsigned short') [-Wint-conversion]
  283 |     return nil;
      |
```

CuriousTommy commented on CoreText/CTFont.m:
```diff
@@ -90,31 +237,129 @@ CGFloat CTFontGetLeading(CTFontRef self) {
+                                        CFIndex maxPositions)
 {
     printf("STUB %s\n", __PRETTY_FUNCTION__);
     return nil;
```
Same here [`CFIndex`](https://developer.apple.com/documentation/corefoundation/cfindex?language=objc)

```
/home/[user]/Downloads/darling2/src/external/cocotron/CoreText/CTFont.m:365:12: error: incompatible pointer to integer conversion returning 'void *' from a function with result type 'CFIndex' (aka 'long') [-Wint-conversion]
  365 |     return nil;
      |
```

CuriousTommy commented on CoreText/CTFont.m:
```diff
@@ -126,27 +371,49 @@ CGFontRef CTFontCopyGraphicsFont(CTFontRef font, CTFontDescriptorRef _Nullable *
+ATSFontRef CTFontGetPlatformFont(CTFontRef font, CTFontDescriptorRef  _Nullable *attributes)
 {
     printf("STUB %s\n", __PRETTY_FUNCTION__);
     return nil;
```
Same here [`ATSFontRef`](https://developer.apple.com/documentation/coretext/atsfontref?language=objc)

```
/home/[user]/Downloads/darling2/src/external/cocotron/CoreText/CTFont.m:385:12: error: incompatible pointer to integer conversion returning 'void *' from a function with result type 'ATSFontRef' (aka 'unsigned int') [-Wint-conversion]
  385 |     return nil;
      |            ^~~
```

CuriousTommy commented on CoreText/CTFont.m:
```diff
@@ -126,27 +371,49 @@ CGFontRef CTFontCopyGraphicsFont(CTFontRef font, CTFontDescriptorRef _Nullable *
+CFTypeID CTFontGetTypeID(void)
+{
+    printf("STUB %s\n", __PRETTY_FUNCTION__);
+    return nil;
```
Same here [`CFTypeID`](https://developer.apple.com/documentation/corefoundation/cftypeid?language=objc)

```
/home/[user]/Downloads/darling2/src/external/cocotron/CoreText/CTFont.m:418:12: error: incompatible pointer to integer conversion returning 'void *' from a function with result type 'CFTypeID' (aka 'unsigned long') [-Wint-conversion]
  418 |     return nil;
      |            ^~~
```

CuriousTommy reviewed (changes requested):
There might be some additional fixes you might need to do, but here is what I found.

CuriousTommy commented on CoreText/CTFont.m:
```diff
@@ -18,17 +18,73 @@ this software and associated documentation files (the "Software"), to deal in
+#import <CoreText/KTFont.h>
 
-const CFStringRef kCTFontStyleNameKey = "CTFontSubFamilyName";
+const CFStringRef kCTFontCopyrightNameKey = CFSTR("CTFontCopyrightName")
```
```
/home/[user]/Downloads/darling2/src/external/cocotron/CoreText/CTFont.m:24:73: error: expected ';' after top level declarator
   24 | const CFStringRef kCTFontCopyrightNameKey = CFSTR("CTFontCopyrightName")
      |                                                                         ^
      | 
```

CuriousTommy commented on CoreText/CTFontDescriptor.m:
```diff
@@ -2,6 +2,29 @@
 
 #include <stdio.h>
 
+const CFStringRef kCTFontURLAttribute = CFSTR("NSCTFontFileURLAttribute");
+const CFStringRef kCTFontNameAttribute = CFSTR("NSFontNameAttribute");
+const CFStringRef kCTFontDisplayNameAttribute = CFSTR("NSFontVisibleNameAttribute");
+const CFStringRef kCTFontFamilyNameAttribute = CFSTR("NSFontFamilyAttribute");
+const CFStringRef kCTFontStyleNameAttribute = CFSTR("NSFontFaceAttribute");
+const CFStringRef kCTFontTraitsAttribute = CFSTR("NSCTFontTraitsAttribute");
+const CFStringRef kCTFontVariationAttribute = CFSTR("NSCTFontVariationAttribute");
+const CFStringRef kCTFontSizeAttribute = CFSTR("NSFontSizeAttribute");
+const CFStringRef kCTFontMatrixAttribute = CFSTR("NSCTFontMatrixAttribute");
+const CFStringRef kCTFontCascadeListAttribute = CFSTR("NSCTFontCascadeListAttribute");
+const CFStringRef kCTFontCharacterSetAttribute = CFSTR("NSCTFontCharacterSetAttribute");
+const CFStringRef kCTFontLanguagesAttribute = CFSTR("NSCTFontLanguagesAttribute");
+const CFStringRef kCTFontBaselineAdjustAttribute = CFSTR("NSCTFontBaselineAdjustAttribute");
+const CFStringRef kCTFontMacintoshEncodingsAttribute = CFSTR("NSCTFontMacintoshEncodingsAttribute");
+const CFStringRef kCTFontFeaturesAttribute = CFSTR("NSCTFontFeaturesAttribute");
+const CFStringRef kCTFontFeatureSettingsAttribute = CFSTR("NSCTFontFeatureSettingsAttribute");
+const CFStringRef kCTFontFixedAdvanceAttribute = CFSTR("NSCTFontFixedAdvanceAttribute");
+const CFStringRef kCTFontOrientationAttribute = CFSTR("NSCTFontOrientationAttribute");
+const CFStringRef kCTFontEnabledAttribute = CFSTR("NSCTFontEnabledAttribute");
+const CFStringRef kCTFontFormatAttribute = CFSTR("NSCTFontFormatAttribute");
+const CFStringRef kCTFontRegistrationScopeAttribute = CFSTR("NSCTFontRegistrationScopeAttribute");
+const CFStringRef kCTFontPriorityAttribute = CFSTR("NSCTFontPriorityAttribute");
```
```
/home/[user]/Downloads/darling2/src/external/cocotron/CoreText/CTFontDescriptor.m:7:41: error: call to undeclared function 'CFSTR'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
    7 | const CFStringRef kCTFontURLAttribute = CFSTR("NSCTFontFileURLAttribute");
      |
```

CuriousTommy reviewed (changes requested):
Just a few more fixes:

https://github.com/darlinghq/darling-cocotron/blob/master/CoreGraphics/include/CoreGraphics/CGContext.h#L22
```
/home/[user]/Downloads/darling2/framework-include/CoreGraphics/CGContext.h:23:32: error: typedef redefinition with different types ('struct CF_BRIDGED_TYPE' vs 'id')
   23 | typedef struct CF_BRIDGED_TYPE(id) O2Context *CGContextRef;
      |                                ^
```

https://github.com/darlinghq/darling-cocotron/blob/master/CoreText/stubs.c#L3-L6
```
[ 45%] Linking C shared library CoreText
ld: warning: OS dylibs should not add rpaths (linker option: -rpath) (Xcode build setting: LD_RUNPATH_SEARCH_PATHS)
duplicate symbol '_CTFontCreateWithFontDescriptor' in:
    CMakeFiles/CoreText.dir/CTFont.m.o
    CMakeFiles/CoreText.dir/stubs.c.o
ld: 1 duplicate symbol for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Lazerbeak12345 commented:
I don't have much experience building darling from source, but I think I'm getting a build error on this branch.

If I'm the only one with the error then I'm likely just missing something.

(when building all of darling, but with this PR checked out in the correct submodule) it succeeds with this:

`cmake -DCOMPONENTS=system -DCMAKE_BUILD_TYPE=Debug ..`

and fails with this

`cmake -DCOMPONENTS=gui -DCMAKE_BUILD_TYPE=Debug ..`

End of the build log when it fails is:

```log
2 warnings generated.
2 warnings generated.
2 warnings generated.
/home/nate/projects/darling/src/external/iokitd/src/IOService.mm:55:13: warning: instance method '-isEqualToString:' not found (return type defaults to 'id') [-Wobjc-method-access]
   55 |                         if ([key isEqualToString: @"IONameMatch"])
      |                                  ^~~~~~~~~~~~~~~
/home/nate/projects/darling/framework-include/Foundation/NSObjCRuntime.h:291:8: note: receiver is instance of class declared here
  291 | @class NSString, Protocol;
      |        ^
2 warnings generated.
3 warnings generated.
[ 92%] Built target Security_i386
2 warnings generated.
2 warnings generated.
[ 92%] Linking CXX shared library HIServices
ld: warning: static initializer found in 'CMakeFiles/HIServices.dir/src/HIServices.c.o'. Use -no_inits to make this an error.  Use -no_warn_inits to suppress warning
ld: warning: static initializer found in 'CMakeFiles/HIServices.dir/src/HIServices.c.o'. Use -no_inits to make this an error.  Use -no_warn_inits to suppress warning
[ 92%] Built target HIServices
2 warnings generated.
2 warnings generated.
[ 92%] Linking CXX executable iokitd
[ 92%] Built target iokitd
make: *** [Makefile:136: all] Error 2
```

The latter _does_ succeed without the changes from this branch.

CuriousTommy commented:
That to be expected, there are some fixes that jlucfarias needs to do before I can merge this.

Lazerbeak12345 commented:
yep! I was just hoping to give it a spin, see if I notice anything, if it works etc.

Lazerbeak12345 commented:
If I apply this patch to this branch, the branch compiles successfully, and my test program prints this near the end of it's logs (it crashes because of stubs, instead of missing symbols, so this is desired).

```patch
diff --git a/CoreGraphics/include/CoreGraphics/CGContext.h b/CoreGraphics/include/CoreGraphics/CGContext.h
index 12a110e5..973670ba 100644
--- a/CoreGraphics/include/CoreGraphics/CGContext.h
+++ b/CoreGraphics/include/CoreGraphics/CGContext.h
@@ -18,6 +18,7 @@ AN ACTION OF CONTRACT,TORT OR OTHERWISE,ARISING FROM,OUT OF OR IN CONNECTION
 WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. */

 #import <CoreGraphics/CoreGraphicsExport.h>
+#import <CoreFoundation/CFBase.h>

 typedef struct CF_BRIDGED_TYPE(id) O2Context *CGContextRef;
```

Here's the (successful) stubs in-use, logged from the software:

```log
STUB CTFontCollectionRef CTFontCollectionCreateFromAvailableFonts(CFDictionaryRef)
STUB CFArrayRef CTFontCollectionCreateMatchingFontDescriptors(CTFontCollectionRef)
```

This patch is needed because the dependency was removed from `CoreText/include/CoreText/CTFontDescriptor.h`, and is next needed in `CoreGraphics/include/CoreGraphics/CGContext.h`

Thanks so much for working on this!

jlucfarias commented:
Sorry for the late response. Many thanks, @Lazerbeak12345. I was unable to search the solution for the error that CuriousTommy noticed and your patch worked like a charm. I've posted and made you as co-author because the change was your, not mine. @CuriousTommy, please review again

CuriousTommy commented:
I was able to build without any build failures occurring. My only final request is that your combine your commits. After that, I'll merge it in.

jlucfarias commented:
@CuriousTommy done

CuriousTommy commented:
Thank you for your contribution!
